flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java

```diff
@@ -201,33 +224,128 @@ private static void visitFactSide(
+                                input, joinKeys, dimSide, dimSideJoinKey, factSideFactors, join));
+            }
+            return union.copy(union.getTraitSet(), newInputs, union.all);
+        } else if (rel instanceof BatchPhysicalGroupAggregateBase) {
+            BatchPhysicalGroupAggregateBase agg = (BatchPhysicalGroupAggregateBase) rel;
+            List<RelNode> newInputs = new ArrayList<>();
+            for (RelNode input : agg.getInputs()) {
+                newInputs.add(
+                        convertDppFactSide(
+                                input,
+                                getInputIndices(agg, input, joinKeys),
+                                dimSide,
+                                dimSideJoinKey,
+                                factSideFactors,
+                                join));
+            }
+
+            return agg.copy(agg.getTraitSet(), newInputs);
         }
+
+        return rel;
     }
 
-    public static List<String> getSuitableDynamicFilteringFieldsInFactSide(
-            DynamicTableSource tableSource, List<String> candidateFields) {
-        List<String> acceptedFilterFields =
-                ((SupportsDynamicFiltering) tableSource).listAcceptedFilterFields();
-        if (acceptedFilterFields == null || acceptedFilterFields.isEmpty()) {
-            return new ArrayList<>();
+    private static boolean isSuitableJoin(Join join) {
+        // Now dynamic partition pruning supports left/right join, inner and semi
+        // join. but now semi
+        // join can not join reorder.
+        if (join.getJoinType() != JoinRelType.INNER
+                && join.getJoinType() != JoinRelType.SEMI
+                && join.getJoinType() != JoinRelType.LEFT
+                && join.getJoinType() != JoinRelType.RIGHT) {
+            return false;
         }
 
-        List<String> suitableFields = new ArrayList<>();
-        // If candidateField not in acceptedFilterFields means dpp rule will not be matched,
-        // because we can not prune any partitions according to non-accepted filter fields
-        // provided by partition table source.
-        for (String candidateField : candidateFields) {
-            if (acceptedFilterFields.contains(candidateField)) {
-                suitableFields.add(candidateField);
+        JoinInfo joinInfo = join.analyzeCondition();
+        return !joinInfo.leftKeys.isEmpty();
+    }
+
+    private static ImmutableIntList getInputIndices(
+            BatchPhysicalGroupAggregateBase agg, RelNode aggInput, ImmutableIntList joinKeys) {
+        int[] indexMap = new int[aggInput.getRowType().getFieldCount()];
+        int[] grouping = agg.grouping();
+        if (grouping.length == 0) {
+            return joinKeys;
+        }
+        int beginIndex = grouping[0] - 1;
+        for (int i = 0; i < indexMap.length; i++) {
+            indexMap[i] = i;
+        }
+
+        System.arraycopy(grouping, 0, indexMap, 0, grouping.length);
+        if (beginIndex >= 0) {
```

Comment:
why we check `beginIndex >= 0` ?

Reply:
> why we check `beginIndex >= 0` ?

`beginIndex >= 0` is the common agg case, which the grouping key is placed in the front of the input index index, and group function is placed later. So `int beginIndex = grouping[0] - 1` will return -1 in this common case.

Reply:
what if the group keys are [0, 2, 5]

Reply:
Done!